Adopt SafeTextMarkerData in more places

cdumez opened the pull request:
#### c0cc24bd0684971e90b5a3dd82e3c6fb03495b02
<pre>
Adopt SafeTextMarkerData in more places
<a href="https://bugs.webkit.org/show_bug.cgi?id=274035">https://bugs.webkit.org/show_bug.cgi?id=274035</a>
<a href="https://rdar.apple.com/127935353">rdar://127935353</a>

Reviewed by Darin Adler.

Rename TextMarkerData to RawTextMarkerData since it contains
raw pointers. Rename SafeTextMarkerData to TextMarkerData and
adopt as much as possible in the code base.

RawTextMarkerData is now only used to convert to and from
bytes.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::visiblePositionForTextMarkerData):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::setNodeIfNeeded const):
(WebCore::AXTextMarker::operator CharacterOffset const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::treeID const):
(WebCore::AXTextMarker::objectID const):
(WebCore::AXTextMarker::node const):

Canonical link: <a href="https://commits.webkit.org/278666@main">https://commits.webkit.org/278666@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96da5463a75e731cb99754bf2e6fd95ec92d3749

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51260 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30564 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3601 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54517 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1950 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53563 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36881 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1627 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41745 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53359 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/28214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44190 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22863 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/25539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1449 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1522 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56113 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26374 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1420 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49145 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27618 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44255 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48297 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28506 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7454 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27352 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -149,18 +143,18 @@ AXTextMarker::operator CharacterOffset() const
         setNodeIfNeeded();
 
-    CharacterOffset result((RefPtr { m_data.node }).get(), m_data.characterStart, m_data.characterOffset);
+    CharacterOffset result((RefPtr { m_data.node.get() }).get(), m_data.characterStart, m_data.characterOffset);
```
I think this expression might have an unneeded extra set of parentheses.

darinadler commented on Source/WebCore/accessibility/AXTextMarker.h:
```diff
@@ -89,9 +89,6 @@ struct TextMarkerData {
-    TextMarkerData(AXObjectCache&, Node*, const VisiblePosition&, int charStart = 0, int charOffset = 0, bool ignoredParam = false);
-    TextMarkerData(AXObjectCache&, const CharacterOffset&, bool ignoredParam = false);
-
     AXID axTreeID() const
```
Are this function and `axObjectID` both still called?

cdumez commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -916,4 +910,15 @@ std::partial_ordering AXTextMarker::partialOrderByTraversal(const AXTextMarker&
+    WeakPtr cache = AXTreeStore<AXObjectCache>::axObjectCacheForID(axTreeID);
+    // Make sure the node is not stale before storing it in a WeakPtr.
+    Node* validNode = nullptr;
+    if (node && cache && cache->isNodeInUse(*node))
```
To address the failures, I had to add this `isNodeInUse()` check to make sure the `node` pointer is not stale after deserializing, before storing in a WeakPtr. `AXObjectCache::isInUse()` is how accessibility code currently does stale node detection.

cdumez committed:
Adopt SafeTextMarkerData in more places
https://bugs.webkit.org/show_bug.cgi?id=274035
rdar://127935353

Reviewed by Darin Adler.

Rename TextMarkerData to RawTextMarkerData since it contains
raw pointers. Rename SafeTextMarkerData to TextMarkerData and
adopt as much as possible in the code base.

RawTextMarkerData is now only used to convert to and from
bytes.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::visiblePositionForTextMarkerData):
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::setNodeIfNeeded const):
(WebCore::AXTextMarker::operator CharacterOffset const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::AXTextMarker):
(WebCore::AXTextMarker::treeID const):
(WebCore::AXTextMarker::objectID const):
(WebCore::AXTextMarker::node const):

Canonical link: https://commits.webkit.org/278666@main
